cpp/src/aztec3/circuits/rollup/base/native_base_rollup_circuit.cpp

```diff
@@ -275,6 +277,92 @@ void perform_historical_contract_data_tree_membership_checks(BaseRollupInputs ba
         check_membership(leaf, historic_root_witness.leaf_index, historic_root_witness.sibling_path, historic_root);
     }
 }
+
+/**
+ * @brief Check non membership of each of the generated nullifiers in the current tree
+ *
+ * @returns The end nullifier tree root
+ */
+AppendOnlySnapshot check_nullifier_tree_non_membership(BaseRollupInputs baseRollupInputs,
+                                                       NT::fr nullifier_tree_subtree_root)
+{
+    // LADIES AND GENTLEMEN The P L A N ( is simple )
+    // 1. Get the previous nullifier set setup
+    // 2. Check for the first added nullifier that it doesnt exist
+    // 3. Update the nullifier set
+    // 4. Calculate a new root with the sibling path
+    // 5. Use that for the next nullifier check.
+    // 6. Iterate for all of em
+    // 7. le bosh (profit)
+
+    // This will update on each iteration
+    auto previous_nullifier_tree_root = baseRollupInputs.start_nullifier_tree_snapshot.root;
+
+    // This will increase with every insertion
+    auto new_index = baseRollupInputs.start_nullifier_tree_snapshot.next_available_leaf_index;
+
+    for (size_t i = 0; i < 2; i++) {
+        auto new_nullifiers = baseRollupInputs.kernel_data[i].public_inputs.end.new_nullifiers;
+        for (size_t j = 0; j < 4; j++) {
+
+            // Witness containing index and path
+            auto witness = baseRollupInputs.low_nullifier_membership_witness[i * 4 + j];
```

Comment:
nit: probably create a var out of `4i+j`
Also replace 4 with the prev constant?